### PR TITLE
convert ItPodsRestart#testServerPodsRestartByChangingResource to JUnit5

### PR DESCRIPTION
**Pull request by xiancao**
Convert ItPodsRestart#testServerPodsRestartByChangingResource to JUnit5 test framework.

The use case covered in testServerPodsRestartByChangingResource:

- Add/Modify server pod resources by patching the domain custom resource.
- Verify all pods were restarted and back to be ready.
- The resources to add: resources: limits: cpu: "1", resources: requests: cpu: "0.5"
- Test fails if any server pod was not restarted and back to be ready or the compute resources in the patched domain custom resource do not match the values we planed to add or modify.

### CODE REVIEW

**Comment by xiancao** [comment]
Jenkins test result: https://build.weblogick8s.org:8443/view/all/job/weblogic-kubernetes-operator-model-in-image-tests/692/
https://build.weblogick8s.org:8443/view/all/job/weblogic-kubernetes-operator-kind-new/112/

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsRestart.java [hunk 1]
```diff
@@ -0,0 +1,370 @@
...
+    // get the admin server pod original creation timestamp
+    logger.info("Getting admin server pod original creation timestamp");
+    String adminPodOriginalTimestamp =
+        assertDoesNotThrow(() -> getPodCreationTimestamp(domainNamespace, "", adminServerPodName),
+            String.format("getPodCreationTimestamp failed with ApiException for pod %s in namespace %s",
+                adminServerPodName, domainNamespace));
+
+    // get the managed server pods original creation timestamps
+    logger.info("Getting managed server pods original creation timestamps");
+    List<String> managedServerPodOriginalTimestampList = new ArrayList<>();
```
Comment (rjeberhard) [inline new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsRestart.java:148]: @sankarpn, @anpanigr, @xiancao... Including all of the pending PR's, it seems like we have three different ways to assert that a pod has been restarted. Let's see if we can come up with a consistent pattern.

Comment (xiancao) [reply new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsRestart.java:148]: Can you list the other ways to assert that a pod has been restarted? In TestAssertions.java, there is only one assertion related to pod restart.

Comment (xiancao) [reply new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsRestart.java:148]: from yesterday's meeting discussion, the cluster server will be rolling restarted after the resources change. I will use Sankar's assertion method once it is ready.

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsRestart.java [hunk 2]
```diff
@@ -0,0 +1,371 @@
...
+import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertNotEquals;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+/**
+ * Test pods were restarted after some properties in server pods changed.
+ */
+@TestMethodOrder(MethodOrderer.OrderAnnotation.class)
```
Comment (sankarpn) [inline new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsRestart.java:63]: Does these tests still needs to be in order?

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsRestart.java [hunk 3]
```diff
@@ -0,0 +1,371 @@
...
+@DisplayName("Test pods were restarted after some properties change in server pods changed")
```
Comment (sankarpn) [inline new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsRestart.java:64]:
```suggestion
@DisplayName("Test pods are restarted after some properties change in server pods are changed")
```

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsRestart.java [hunk 4]
```diff
@@ -0,0 +1,371 @@
...
+    // install and verify operator
+    installAndVerifyOperator(opNamespace, domainNamespace);
+
+    // create a basic model in image domain
+    createAndVerifyMiiDomain();
+  }
+
+  /**
+   * Add/Modify server pod resources by patching the domain custom resource.
+   * Verify all pods were restarted and back to be ready.
```
Comment (sankarpn) [inline new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsRestart.java:105]:
```suggestion
   * Verify all pods are restarted and back to ready state.
```

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsRestart.java [hunk 5]
```diff
@@ -0,0 +1,371 @@
...
+   * The resources tested: resources: limits: cpu: "1", resources: requests: cpu: "0.5"
+   * Test fails if any server pod was not restarted and back to be ready or the compute resources in the patched
```
Comment (sankarpn) [inline new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsRestart.java:107]:
```suggestion
   * Test fails if any server pod is not restarted and back to ready state or the compute resources in the patched
```

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsRestart.java [hunk 6]
```diff
@@ -0,0 +1,371 @@
...
+   * domain custom resource do not match the values we planed to add or modify.
```
Comment (sankarpn) [inline new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsRestart.java:108]:
```suggestion
   * domain custom resource do not match the values we planned to add or modify.
```

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsRestart.java [hunk 7]
```diff
@@ -0,0 +1,371 @@
...
+   */
+  @Test
+  @DisplayName("Verify server pods were restarted by changing the resources")
```
Comment (sankarpn) [inline new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsRestart.java:111]:
```suggestion
  @DisplayName("Verify server pods are restarted by changing the resources")
```

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsRestart.java [hunk 8]
```diff
@@ -0,0 +1,371 @@
...
+  public void testServerPodsRestartByChangingResource() {
+
+    // get the original domain resource before update
+    Domain domain1 = assertDoesNotThrow(() -> getDomainCustomResource(domainUid, domainNamespace),
+        String.format("getDomainCustomResource failed with ApiException when tried to get domain %s in namespace %s",
+            domainUid, domainNamespace));
+
+    assertNotNull(domain1, "domain1 is null");
```
Comment (sankarpn) [inline new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsRestart.java:119]:
```suggestion
    assertNotNull(domain1, domain1 + " is null");
```

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsRestart.java [hunk 9]
```diff
@@ -0,0 +1,371 @@
...
+    assertNotNull(domain1.getSpec(), "domain1/spec is null");
```
Comment (sankarpn) [inline new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsRestart.java:120]:
```suggestion
    assertNotNull(domain1.getSpec(), domain1 + "/spec is null");
```

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsRestart.java [hunk 10]
```diff
@@ -0,0 +1,371 @@
...
+    assertNotNull(domain1.getSpec().getServerPod(), "domain1/spec/serverPod is null");
+    assertNotNull(domain1.getSpec().getServerPod().getResources(), "domain1/spec/serverPod/resources is null");
```
Comment (sankarpn) [inline new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsRestart.java:122]: same as above rest of the class

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsRestart.java [hunk 11]
```diff
@@ -0,0 +1,371 @@
...
+
+    // verify the managed server pods were rolling restarted and back to be ready
+    Map<String, String> managedServerPodsWithTimeStamps = new HashMap<>();
+    for (int i = 1; i <= replicaCount; i++) {
+      String managedServerPodName = managedServerPrefix + i;
+      managedServerPodsWithTimeStamps.put(managedServerPodName, managedServerPodOriginalTimestampList.get(i - 1));
+    }
+    logger.info("Verifying rolling restart occurred for managed servers {0} in namespace {1}",
+        managedServerPodsWithTimeStamps.keySet(), domainNamespace);
+    verifyRollingRestartOccurred(managedServerPodsWithTimeStamps, 1, domainNamespace);
```
Comment (sankarpn) [inline new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsRestart.java:195]: This assertion will make sure that the pods - admin server, managed servers are restarted in order and comes to ready state.
So you don't need to check anything after doing addServerPodResources(cpuLimit, cpuRequest)
Please see the example in ItMiiConfigMapOverride test testAddMiiSystemResources on how to use this assertion.

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsRestart.java [hunk 12]
```diff
@@ -0,0 +1,371 @@
...
+    // get new server pod compute resources requests
+    requests = domain1.getSpec().getServerPod().getResources().getRequests();
+    assertNotNull(requests, "domain1/spec/serverPod/resources/requests is null");
+
+    // print out server pod compute resource requests
+    logger.info("New value for server pod compute resource requests:");
+    requests.forEach((key, value) -> logger.info(key + ": " + value.getNumber().toString()));
+
+    // verify the server pod resources requests got updated
+    logger.info("Checking that the server pod resources cpu request was updated correctly");
```
Comment (sankarpn) [inline new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsRestart.java:231]:
```suggestion
    logger.info("Checking that the server pod resources cpu request are updated correctly");
```

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsRestart.java [hunk 13]
```diff
@@ -0,0 +1,371 @@
...
+
+    // check that admin server pod is ready
+    logger.info("Checking that admin server pod {0} is ready in namespace {1}",
+        adminServerPodName, domainNamespace);
+    checkPodReady(adminServerPodName, domainUid, domainNamespace);
+
+    // check that admin service exists in the domain namespace
+    logger.info("Checking that admin service {0} exists in namespace {1}",
+        adminServerPodName, domainNamespace);
+    checkServiceExists(adminServerPodName, domainNamespace);
```
Comment (sankarpn) [inline new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsRestart.java:320]: This has to be done right after checkPodExists(adminServerPodName, domainUid, domainNamespace);

Comment (xiancao) [reply new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsRestart.java:320]: why?

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsRestart.java (hunk 1 above)
Comment (xiancao) [reply new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsRestart.java]: updated the code using verifyRollingRestartOccurred()

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsRestart.java [hunk 14]
```diff
@@ -0,0 +1,371 @@
...
+import static oracle.weblogic.kubernetes.utils.CommonTestUtils.dockerLoginAndPushImageToRegistry;
+import static oracle.weblogic.kubernetes.utils.CommonTestUtils.installAndVerifyOperator;
+import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertNotEquals;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+/**
+ * Test pods were restarted after some properties in server pods changed.
```
[the remark this answers is not in the document]
Comment (xiancao) [reply new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsRestart.java]: fixed

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsRestart.java (hunk 2 above)
Comment (xiancao) [reply new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsRestart.java]: removed the annotation

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsRestart.java (hunk 3 above)
Comment (xiancao) [reply new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsRestart.java]: fixed

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsRestart.java (hunk 4 above)
Comment (xiancao) [reply new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsRestart.java]: fixed

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsRestart.java (hunk 5 above)
Comment (xiancao) [reply new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsRestart.java]: fixed

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsRestart.java (hunk 6 above)
Comment (xiancao) [reply new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsRestart.java]: fixed

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsRestart.java (hunk 7 above)
Comment (xiancao) [reply new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsRestart.java]: fixed

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsRestart.java (hunk 8 above)
Comment (xiancao) [reply new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsRestart.java]: fixed

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsRestart.java (hunk 9 above)
Comment (xiancao) [reply new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsRestart.java]: fixed

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsRestart.java (hunk 10 above)
Comment (xiancao) [reply new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsRestart.java]: fixed

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsRestart.java (hunk 11 above)
Comment (xiancao) [reply new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsRestart.java]: fixed

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsRestart.java (hunk 12 above)
Comment (xiancao) [reply new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsRestart.java]: fixed

**Comment by xiancao** [comment]
Jenkins test result: https://build.weblogick8s.org:8443/job/weblogic-kubernetes-operator-kind-new/183/

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsRestart.java (hunk 13 above)
Comment (sankarpn) [reply new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsRestart.java:305]: The services are created right after the pod is created. If there was a problem in creating the services for admin server the operator will not create managed server pods. But the test code won't check service existence for admin server pod until the managed server are ready but that will never happen. The test will fail with a timedout condition waiting for managed server pod when in fact it should fail for non existence service for admin server

Comment (xiancao) [reply new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsRestart.java:305]: The test checks the admin service existence after the admin server pod is ready, not after the managed server pod is ready.

**Comment by xiancao** [comment]
Jenkins test result: https://build.weblogick8s.org:8443/view/all/job/weblogic-kubernetes-operator-kind-new/187/
https://build.weblogick8s.org:8443/view/all/job/weblogic-kubernetes-operator-model-in-image-tests-10/189/

**Pull request merged**